### ISSUE TITLE
Import broken on master as of commit 1cb37b8bc.

### ISSUE THREAD

**Issue by kfogel** [OpenTechStrategies/smoke-alarm-portal#63]
This is on `master` with HEAD commit 1cb37b8bc, after a freshly dropped and recreated database.  Right now it happens for @kfogel every time, but @trystant can't reproduce it at all -- his imports seem to work fine (except for the #62 errors, which are expected and at least not fatal).

@trystant suggested changing "`force: false`" to "`force: true`" (in models/index.js) before the import, but the error spew happens for @kfogel either way, and no working database results.

    $ node data/import_into_postgres.js
    DEBUG: Before DB Sync
    Unhandled rejection SequelizeDatabaseError: relation "SelectedCounties" does not exist
        at Query.formatError (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:423:14)
        at null.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:104:19)
        at emit (events.js:95:17)
        at Query.handleError (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/query.js:99:8)
        at null.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/client.js:166:26)
        at emit (events.js:95:17)
        at Socket.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/connection.js:109:12)
        at Socket.emit (events.js:95:17)
        at Socket.<anonymous> (_stream_readable.js:765:14)
        at Socket.emit (events.js:92:17)
        at emitReadable_ (_stream_readable.js:427:10)
        at emitReadable (_stream_readable.js:423:5)
        at readableAddChunk (_stream_readable.js:166:9)
        at Socket.Readable.push (_stream_readable.js:128:10)
        at TCP.onread (net.js:529:21)
    Unhandled rejection SequelizeDatabaseError: relation "SelectedCounties" does not exist
        at Query.formatError (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:423:14)
        at null.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:104:19)
        at emit (events.js:95:17)
        at Query.handleError (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/query.js:99:8)
        at null.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/client.js:166:26)
        at emit (events.js:95:17)
        at Socket.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/connection.js:109:12)
        at Socket.emit (events.js:95:17)
        at Socket.<anonymous> (_stream_readable.js:765:14)
        at Socket.emit (events.js:92:17)
        at emitReadable_ (_stream_readable.js:427:10)
        at emitReadable (_stream_readable.js:423:5)
        at readableAddChunk (_stream_readable.js:166:9)
        at Socket.Readable.push (_stream_readable.js:128:10)
        at TCP.onread (net.js:529:21)
    
    ##### We get a ton of those, and then we start getting a ton of these: #####
    
    Unhandled rejection SequelizeDatabaseError: relation "UsAddress" does not exist
        at Query.formatError (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:423:14)
        at null.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:104:19)
        at emit (events.js:95:17)
        at Query.handleError (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/query.js:99:8)
        at null.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/client.js:166:26)
        at emit (events.js:95:17)
        at Socket.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/pg/lib/connection.js:109:12)
        at Socket.emit (events.js:95:17)
        at Socket.<anonymous> (_stream_readable.js:765:14)
        at Socket.emit (events.js:92:17)
        at emitReadable_ (_stream_readable.js:427:10)
        at emitReadable (_stream_readable.js:423:5)
        at readableAddChunk (_stream_readable.js:166:9)
        at Socket.Readable.push (_stream_readable.js:128:10)
        at TCP.onread (net.js:529:21)

**Comment by kfogel** [comment]
After the above, I did `git checkout 58_email_sending`, where HEAD commit is 9a3144b51c, and ran `node data/import_into_postgres.js`, and the import worked (though as expected it still gets the #62 errors).  So it does seem like something in `master` is the cause.

**Comment by kfogel** [comment]
@cecilia-donnelly and I both tried doing the installation & import steps from scratch, with a dropped and recreated empty DB, and, interestingly, both got different errors from each other and from the above.  The errors happened on the import step.

Mine is recorded at http://paste.lisp.org/display/152010, but in case that disappears, here it is:

    $ git branch
      15_contact_preference
      58_email_sending
      drywall_integration
      email_demo
      hacky-layout-experiment
      issue-43-experimental
    * master
      rc-regions-by-county-rename
    $ git status
    On branch master
    Your branch is up-to-date with 'origin/master'.
    Untracked files:
      (use "git add <file>..." to include in what will be committed)
    
    	config.js
    
    nothing added to commit but untracked files present (use "git add" to track)
    $ cat config.js
    'use strict';
    
    exports.port = process.env.PORT || 3000;
    exports.companyName = 'Open Tech Strategies, LLC.';
    exports.projectName = 'Red Cross Smoke Alarm Installation Request Portal';
    exports.systemEmail = 'jrandom@example.com';
    
    // Make up your own key here.  More background:
    //
    // I believe the cryptoKey is used to server-side sign session data
    // that is then sent in a cookie to the client -- the client never
    // knows the key, so the client therefore cannot tamper with any of
    // the data in the cookie, because if it did, the digital signature on
    // the cookie would be invalid.
    //
    // These two pages have more information.  In the first page, I think
    // where they say "encrypt" they really mean "sign"; the second page
    // explains the situation pretty straightforwardly, but it helps to
    // have read the first page for context:
    // 
    //   http://stackoverflow.com/questions/18565512/importance-of-session-secret-key-in-express-web-framework
    //   http://stackoverflow.com/questions/6719036/why-cherrypy-session-does-not-require-a-secret-key
    exports.cryptoKey = 'k3yb0ardc4t';
    
    exports.loginAttempts = {
      forIp: 50,
      forIpAndUser: 7,
      logExpiration: '20m'
    };
    exports.requireAccountVerification = false;
    exports.smtp = {
      from: {
        name: process.env.SMTP_FROM_NAME || exports.projectName +' Website',
        address: process.env.SMTP_FROM_ADDRESS || 'jrandom@example.com'
      },
      credentials: {
        user: process.env.SMTP_USERNAME || 'jrandom@example.com',
        password: process.env.SMTP_PASSWORD || '',
        host: process.env.SMTP_HOST || 'smtp.gmail.com',
        ssl: true
      }
    };
    exports.oauth = {
      twitter: {
        key: process.env.TWITTER_OAUTH_KEY || '',
        secret: process.env.TWITTER_OAUTH_SECRET || ''
      },
      facebook: {
        key: process.env.FACEBOOK_OAUTH_KEY || '',
        secret: process.env.FACEBOOK_OAUTH_SECRET || ''
      },
      github: {
        key: process.env.GITHUB_OAUTH_KEY || '',
        secret: process.env.GITHUB_OAUTH_SECRET || ''
      },
      google: {
        key: process.env.GOOGLE_OAUTH_KEY || '',
        secret: process.env.GOOGLE_OAUTH_SECRET || ''
      },
      tumblr: {
        key: process.env.TUMBLR_OAUTH_KEY || '',
        secret: process.env.TUMBLR_OAUTH_SECRET || ''
      }
    };
    $ node data/import_into_postgres.js
    
    /home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/models/npm-debug.log:1
    (function (exports, require, module, __filename, __dirname) { 0 info it worked
                                                                    ^^^^
    SyntaxError: Unexpected identifier
        at Module._compile (module.js:439:25)
        at Object.Module._extensions..js (module.js:474:10)
        at Module.load (module.js:356:32)
        at Function.Module._load (module.js:312:12)
        at Module.require (module.js:364:17)
        at require (module.js:380:17)
        at Sequelize.import (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/node_modules/sequelize/lib/sequelize.js:599:61)
        at options.force (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/models/index.js:19:40)
        at Array.forEach (native)
        at Object.<anonymous> (/home/kfogel/private/work/ots/client/red-cross/r/smoke-alarm-portal/models/index.js:18:6)
        at Module._compile (module.js:456:26)
        at Object.Module._extensions..js (module.js:474:10)
        at Module.load (module.js:356:32)
        at Function.Module._load (module.js:312:12)
        at Module.require (module.js:364:17)
        at require (module.js:380:17)
    $ 

And @cecilia-donnelly's is at http://pamrel.lu/915d3/, but again, in case that disappears, here it is:

    Executing (default): INSERT INTO "SelectedCounties" ("id","region","state","county","updatedAt","createdAt") VALUES (DEFAULT,'Iowa','Iowa','Allamakee','2015-07-21 00:02:12.651 +00:00','2015-07-21 00:02:12.651 +00:00') RETURNING *;
    
    Unhandled rejection SequelizeDatabaseError: relation "SelectedCounties" does not exist
        at Query.formatError (/home/cdonnelly/OTS/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:427:14)
        at null.<anonymous> (/home/cdonnelly/OTS/smoke-alarm-portal/node_modules/sequelize/lib/dialects/postgres/query.js:108:19)
        at emit (events.js:95:17)
        at Query.handleError (/home/cdonnelly/OTS/smoke-alarm-portal/node_modules/pg/lib/query.js:99:8)
        at null.<anonymous> (/home/cdonnelly/OTS/smoke-alarm-portal/node_modules/pg/lib/client.js:166:26)
        at emit (events.js:95:17)
        at Socket.<anonymous> (/home/cdonnelly/OTS/smoke-alarm-portal/node_modules/pg/lib/connection.js:109:12)
        at Socket.emit (events.js:95:17)
        at Socket.<anonymous> (_stream_readable.js:765:14)
        at Socket.emit (events.js:92:17)
        at emitReadable_ (_stream_readable.js:427:10)
        at emitReadable (_stream_readable.js:423:5)
        at readableAddChunk (_stream_readable.js:166:9)
        at Socket.Readable.push (_stream_readable.js:128:10)
        at TCP.onread (net.js:529:21)

**Comment by trystant** [comment]
I'm running into this error now. I believe that the issue is related to the async operation not syncing up with the Promise returned by the Sequelize query and the results. I am going to refit the import to use promises only and see how that pans out.

**Comment by kfogel** [comment]
Why not just convert the json to sql and import directly?

**Comment by trystant** [comment]
We will have to learn the lay of the land sooner or later. That pushes
this back to another ticket at another time. I'm fine with that
decision if we want to keep it simple for now, however.

**Comment by trystant** [comment]
For the record, I'm looking at this [script](https://github.com/OpenTechStrategies/smoke-alarm-portal/blob/email_demo/data/determine_max_strings) as mentioned as a guideline for creating SQL out of JSON, although I am not sure what aspect of the parsing library carries over to this import script.

**Comment by trystant** [comment]
The import's been updated to create SQL from JSON, as discussed. The SQL has been confirmed to import into Postgresql successfully. I have some documentation left to do, but the updated import obviates many past import issues.

**Comment by kfogel** [comment]
Thanks!  Just for the record, commit fe99582e3 (and taking a look now).

**Comment by kfogel** [comment]
See commit 398428928 on the new `data_import_fixes` branch; worth reading the commit message and code changes in detail.

(Hmm, I probably should have mentioned this issue and/or issue #51 in the commit message, as we should have with the other related commits, but no big deal.  Let's try to remember it for next time.)

Mark, we should probably consult tomorrow; I hope you'll be online.  I have a better understanding now than I did before of what Sequelize was doing for us (based on the files in `models/`) in the old direct-from-JSON method.  I *think* that after commit 398428928 we must be pretty close, but I could be wrong.  It may -- _may_ -- don't kill me -- make sense to go back to the old way, if it turns out that there are a zillion little things we were getting from Sequelize that we're not getting this other way.  I just don't know.  We should discuss, once we're both online, but please don't wait on that discussion.  Try whatever means seem best to you to get us over this data import hump and to development of the admin area.

**Comment by kfogel** [comment]
http://www.neilconway.org/docs/sequences/ might point the way to getting the sequence setup right (I'm pretty sure I got it wrong, in `data/create_static_tables.sql`).

**Comment by trystant** [comment]
Good Morning Karl,

I think that we are close.

I also think that we can sync up better; looking at that commit (3984289),
I am confused as to a couple of things; data/create_static_tables.sql in
particular.

As of now, there's a script which creates SQL that can be imported into
Postgres successfully. The only remaining steps to overcome would be to
clean up the script execution (putting the SQL files in a proper directory)
and running the command line scripts to import.

I don't think there's a hump regarding an import at this point. I can write
steps to document how to install from SQL, the documentation is not done.

I did not report that the updated script was not a complete script; I had
planned on updating that this morning, but I see that it's already been
modified. I should have reported my next steps in the ticket or the commit.

I think that changing the table name from UsAddress to UsAddresses merits
raising the question; do we want table names to be singular or plural?

Generally speaking, there's a lot that we do gain from Sequelize, and if I
could have replayed yesterday, I believe the time would have been better
spent implementing it.

I can jump on IRC now, but it's 3:19 AM your time. I'll check in quickly
now, however.

On Wed, Jul 22, 2015 at 5:57 AM, Karl Fogel <notifications@github.com>
wrote:

> See commit 3984289
> <https://github.com/OpenTechStrategies/smoke-alarm-portal/commit/39842892840f319ed7c83deeec136056e6d0d2dd>
> on the new data_import_fixes branch; worth reading the commit message and
> code changes in detail.
>
> (Hmm, I probably should have mentioned this issue and/or issue #51
> <https://github.com/OpenTechStrategies/smoke-alarm-portal/issues/51> in
> the commit message, as we should have with the other related commits, but
> no big deal. Let's try to remember it for next time.)
>
> Mark, we should probably consult tomorrow; I hope you'll be online. I have
> a better understanding now than I did before of what Sequelize was doing
> for us (based on the files in models/) in the old direct-from-JSON
> method. I *think* that after commit 3984289
> <https://github.com/OpenTechStrategies/smoke-alarm-portal/commit/39842892840f319ed7c83deeec136056e6d0d2dd>
> we must be pretty close, but I could be wrong. It may -- *may* -- don't
> kill me -- make sense to go back to the old way, if it turns out that there
> are a zillion little things we were getting from Sequelize that we're not
> getting this other way. I just don't know. We should discuss, once we're
> both online, but please don't wait on that discussion. Try whatever means
> seem best to you to get us over this data import hump and to development of
> the admin area.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenTechStrategies/smoke-alarm-portal/issues/63#issuecomment-123647186>
> .
>



-- 

Mark Nyonmark@grandkru.com

blog.grandkru.com

github.com/trystant

**Comment by trystant** [comment]
This commit f3bdcb2 has an import that successfully uses Sequelize to save data into the database. This should close this ticket, but I'm waiting for someone else to confirm the behavior.

**Comment by cecilia-donnelly** [comment]
@kfogel and I were both able to run the import on master yesterday, after commit f3bdcb2.
